java/ql/src/semmle/code/java/dataflow/internal/DataFlowImpl.qll

```diff
@@ -878,23 +893,28 @@ private module Stage2 {
 
   pragma[nomagic]
   private predicate fwdFlowIn(
-    DataFlowCall call, ParameterNode p, Cc cc, ApOption argAp, Ap ap, Configuration config
+    DataFlowCall call, ParameterNode p, Cc outercc, Cc innercc, ApOption argAp, Ap ap,
+    Configuration config
   ) {
     exists(ArgumentNode arg, boolean allowsFieldFlow |
-      fwdFlow(arg, cc, argAp, ap, config) and
-      flowIntoCallNodeCand1(call, arg, p, allowsFieldFlow, config)
+      fwdFlow(arg, outercc, argAp, ap, config) and
+      flowIntoCallNodeCand1(call, arg, p, allowsFieldFlow, config) and
+      innercc = getCallContextCall(call, p.getEnclosingCallable(), outercc)
     |
       ap instanceof ApNil or allowsFieldFlow = true
     )
   }
 
   pragma[nomagic]
   private predicate fwdFlowOut(
-    DataFlowCall call, Node node, Cc cc, ApOption argAp, Ap ap, Configuration config
+    DataFlowCall call, Node node, Cc innercc, Cc ccOut, ApOption argAp, Ap ap, Configuration config
```

Comment:
I'm a little bit in doubt about that rename, since `ccOut` isn't necessarily the outer call context we end up with if we are in the flow-through case, because then we are re-establishing the outer call context, rather than using the return-based call context that's calculated here. Maybe this predicate should just have qldoc to clarify this.

Reply:
👍